docs/adr/0006-2021-05-26-authorisation.md

```diff
@@ -0,0 +1,92 @@
+
+(Note that Azure AD is a distinct product from Azure AD B2C and on-prem Active Directory for Windows Server. More info
+[here](https://www.predicagroup.com/blog/azure-ad-b2b-b2c-puzzled-out/).)
+
+## Decision
+
+_What is the change that we're proposing and/or doing?_
+
+We will implement a blended authorisation architecture for the Beacons domain using Azure B2C for public-facing
+accounts, Azure AD for internal accounts with role-based access control, and application code.
+
+1.  End-users (members of the public: beacon owners, account holders) will authenticate using Azure B2C to
+    access the consumer-facing [Web App](https://github.com/mcagov/beacons-webapp).
+2.  The Web App will authenticate with Azure AD as a [confidential client application](https://docs.microsoft.com/en-us/azure/healthcare-apis/fhir/register-confidential-azure-ad-client-app) to access the
+    Beacons [Service API](https://github.com/mcagov/beacons-service) with scope to perform CRUD operations on all
+    beacon records.
+3.  Application code in the Web App will restrict the permissions of the current Beacon Owner / public account
+    holder to operate only on beacon records within their scope.
+4.  Maritime & Coastguard Agency internal users will authenticate using Azure AD to access the Beacons [Back Office
+    Admin Tool](https://github.com/mcagov/beacons-backoffice). The Maritime & Coastguard Agency's production tenant
+    will be used to allow Single Sign-On. Using Azure AD will allow role-based access control of back office
+    operations, giving the ability to make Search & Rescue users 'read only' while allowing Beacon Registration Team
+    users to add flags, comments and other metadata.
+
+![Beacons authorisation diagram](assets/beacons-auth-diagram.png).
+
+## Key Benefits
+
+_List the main benefits of this implementation._
+
+- Commonality with other Maritime & Coastguard Agency services.
+
+- Single-sign on for internal users.
+
+- The lowest implementation effort (see alternatives).
+
+## Key Drawbacks
+
+_List any significant drawbacks from this implementation._
+
+- Setting the consumer-facing web app as a confidential client which authenticates with the Beacons Service API with
+  an API key foregoes the opportunity to implement delegated access, where the OAuth 2.0 protocol manages end-users
+  directly authorising access to their beacons in the API. Knowledge of which user is performing an operation via
+  the Web App must be implemented in the application.
+  - The confidential client key held by the Web App represents a larger blast radius in the event of a compromised
+    credential than the alternative delegation model (see alternative).
+
+## Alternatives
+
+_Summarise the main alternatives that have been considered and the reasons why they have not been chosen. Use no
+more than one paragraph per alternative._
+
+1.  **Full OAuth 2.0 delegation model**. The beacon owner / account holder is made a true OAuth 2.0 resource owner that
```

Comment:
I would add to this that this current option isn't possible due to B2C not supporting on behalf of user flows

Reply:
Ah I think this option is possible with B2C? 

The [on-behalf-of](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow) user flow would be required if there was a protected resource at an API behind our API, so the chain would go `resource owner (user) > client (webapp) > intermediary (registrations API) -> protected resource (another API)`.

![image](https://user-images.githubusercontent.com/54271433/120609513-23ab8780-c44a-11eb-958d-00828e00e463.png)

What I mean by the full OAuth 2.0 delegation model is the standard authorization code grant type (`resource owner (user) -> client (webapp) -> protected resource (registrations API)`).  

I _hope_ I've got this right 😅. I've added a couple of sentences to the option to clarify.